### PR TITLE
feat: verify contracts on deploy

### LINKED ISSUE #47: Verify contracts on deploy
Registering an upkeep via the web app requires the contract to be verified.

### PR DESCRIPTION
**Pull request by petia-latinova**
resolves #47

### CODE REVIEW

**Review by ivaylonikolov7 — COMMENTED** [review]
Is there a reason we need the try-catch?

**Comment by imollov** [comment]
@petia-latinova harvester template should also be verified.

File: batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts [hunk 1]
```diff
@@ -47,22 +47,51 @@ const func: DeployFunction = async function ({
...
-    ],
+    args: args,
     log: true,
+    waitConfirmations:
+      network.name !== "hardhat" && network.name !== "localhost"
+        ? 6
+        : undefined,
   });
+
+  if (network.name !== "hardhat" && network.name !== "localhost") {
```
Comment (imollov) [inline batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts:73]: Why not use `developmentChains` like on line 26?

**Comment by imollov** [comment]
> > Is there a reason we need the try-catch?
> 
> If the contract is already verified it throws error in the console. With the try-catch I console.log it and make it shorter. Maybe it's better then error.

@petia-latinova Although I see your point, I'm not sure the added code complexity is worth the benefit of having shorter error.

**Comment by petia-latinova** [comment]
When you deploy the contract in batch-nft-reveal there is the following  warning and I think will be nice if is possible to be fixed:

Warning: Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable.
      --> contracts/NFTCollection.sol:276:37:
        returns (bool upkeepNeeded, bytes memory)

File: vault-harvester/scripts/deploy.ts [hunk 2]
```diff
@@ -58,17 +58,40 @@ const deployContract = async (
   contractFactory: ContractFactory,
   constructorArgs: any[]
 ): Promise<{ contract: string; impl: string }> => {
+  if (network.name !== "matic") {
```
Comment (petia-latinova) [inline vault-harvester/scripts/deploy.ts:61]: Maybe I need to add here and the mainnet for etherium so not to deploy MockKeeperRegistry on mainnet?

File: vault-harvester/scripts/deploy.ts [hunk 3]
```diff
@@ -58,17 +58,40 @@ const deployContract = async (
...
   const contract = await ethers.getContractAt(name, deployTx.address);
   ret.contract = contract.address;
 
+  if (network.name !== "hardhat" && network.name !== "localhost") {
+    console.log("Waiting to verify...");
+    await deployTx.deployTransaction.wait(6);
+    console.log(`Verifying ${config.harvester.contractName}...`);
+    await run("verify:verify", {
+      address: deployTx.address,
+      contract: `contracts/test/MockHarvester.sol:${config.harvester.contractName}`,
```
Comment (ivaylonikolov7) [inline vault-harvester/scripts/deploy.ts:90]: Why `contracts/test/MockHarvester.sol`? Isn't this part supposed to be not for deploying test contracts?

File: batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts [hunk 4]
```diff
@@ -47,22 +48,34 @@ const func: DeployFunction = async function ({
...
+    NFT_MINT_COST,
+    NFT_REVEAL_BATCH_SIZE,
+    NFT_REVEAL_INTERVAL,
+    vrfCoordinatorV2Address,
+    vrfSubscriptionId,
+    vrfGasLane,
+    vrfCallbackGasLimit,
+  ];
+
+  const result = await deploy("NFTCollection", {
```
Comment (ivaylonikolov7) [inline batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts:64]: Maybe extract this string "NFTCollection" to another variable similar to deployment scripts template-nft-collection-erc721?

**Comment by imollov** [comment]
@petia-latinova `deployContract` is supposed to be a generic deploy function. All the changes should be a separate functions called from outside.

File: vault-harvester/scripts/deploy.ts (hunk 3 above)
Comment (petia-latinova) [reply vault-harvester/scripts/deploy.ts:90]: What else you suggest can be deployed in this project?

File: vault-harvester/scripts/deploy.ts (hunk 2 above)
Comment (petia-latinova) [reply vault-harvester/scripts/deploy.ts:61]: This deploy script works only for matic mainnet, because the address of the keeper contract is hardcoded in the beginning. it'll fail for any other network if the mock keeper is not deployed. It wasn't possible to dpeloy it with that script on testnets. There is the same address for etherium mainnet that's why I ask that question. 

But if somebody change the address the if statement won't be valid. If somebody have suggestions how this can be done better is welcome


I don't see how the user can deploy contract without doing other changes so maybe is better without deploy script or I can remove the code for testnetworks and add the verify code without testing it. I think I already tested it actually but will be possible to be deployed only on poligon mainnet and maybe etherium mainnet.

File: vault-harvester/scripts/deploy.ts [hunk 5]
```diff
@@ -58,17 +58,40 @@ const deployContract = async (
...
 ): Promise<{ contract: string; impl: string }> => {
+  if (network.name !== "matic") {
+    console.log("Deploying: MockKeeperRegistry");
+    const MockKeeperRegistryFactory = await ethers.getContractFactory(
+      "MockKeeperRegistry"
+    );
+    const mockKeeperRegistry = await MockKeeperRegistryFactory.deploy();
+    await mockKeeperRegistry.deployed();
+    console.log(`MockKeeperRegistry: ${mockKeeperRegistry.address}`);
+    constructorArgs[0] = mockKeeperRegistry.address;
```
Comment (imollov) [inline vault-harvester/scripts/deploy.ts:69]: Deploying this mock is not necessary because the deploy script requires a deployed `KeeperRegistry` address. It's described in the readme that testing the deployment locally should be done on a forked network - https://github.com/hackbg/chainlink-keepers-templates/tree/main/vault-harvester#deploy.

File: vault-harvester/scripts/deploy.ts [hunk 6]
```diff
@@ -63,12 +63,24 @@ const deployContract = async (
...
+    console.log("Waiting to verify...");
+    await deployTx.deployTransaction.wait(6);
+    console.log(`Verifying ${config.harvester.contractName}...`);
+    await run("verify:verify", {
+      address: deployTx.address,
+      contract: `contracts/test/${config.harvester.contractName}.sol:${config.harvester.contractName}`,
+      constructorArguments: constructorArgs,
+    });
+  }
+
```
Comment (imollov) [inline vault-harvester/scripts/deploy.ts:83]: Can we move this to `deployHarvester` because `deployContract` is a generic function which only works with its parameters as input.

File: batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts [hunk 7]
```diff
@@ -16,14 +16,17 @@ const func: DeployFunction = async function ({
   getNamedAccounts,
   getChainId,
 }: HardhatRuntimeEnvironment) {
+  const contractName = "NFTCollection";
+
```
Comment (imollov) [inline batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts:20]:
```suggestion
```

File: batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts [hunk 8]
```diff
@@ -47,22 +50,34 @@ const func: DeployFunction = async function ({
...
-      NFT_REVEAL_BATCH_SIZE,
-      NFT_REVEAL_INTERVAL,
-      vrfCoordinatorV2Address,
-      vrfSubscriptionId,
-      vrfGasLane,
-      vrfCallbackGasLimit,
-    ],
+    args: args,
     log: true,
+    waitConfirmations: !isDevChain ? 6 : undefined,
```
Comment (imollov) [inline batch-nft-reveal/deploy/02_Deploy_NFTCollection.ts:70]:
```suggestion
    waitConfirmations: isDevChain ? 0 : 6,
```

File: vault-harvester/hardhat.config.ts [hunk 9]
```diff
@@ -33,12 +33,23 @@ const config: HardhatUserConfig = {
...
+      accounts:
+        process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
+    },
+    rinkeby: {
+      url: process.env.RINKEBY_URL || "",
+      accounts:
+        process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
+    },
+    mumbai: {
+      url: process.env.MUMBAI_URL || "",
```
Comment (imollov) [inline vault-harvester/hardhat.config.ts:52]: Please make these changes in a separate commit.

**Review by ivaylonikolov7 — APPROVED** [review]

File: vault-harvester/scripts/deploy.ts [hunk 10]
```diff
@@ -46,30 +46,37 @@ const deployHarvester = async () => {
...
-    contract: "",
-    impl: "",
-  };
+
   const deployTx = await contractFactory.deploy(...constructorArgs);
   await deployTx.deployed();
   console.log(`${name}: ${deployTx.address}`);
-  const contract = await ethers.getContractAt(name, deployTx.address);
-  ret.contract = contract.address;
+  await ethers.getContractAt(name, deployTx.address);
```
Comment (imollov) [inline vault-harvester/scripts/deploy.ts:77]: What is this used for?

**Review by imollov — APPROVED** [review]
